documentation/design/proposals/authn_authz_for_1.0.adoc

```diff
@@ -0,0 +1,329 @@
+The Access Control Server will perform a simple bind against the LDAP in order to determine the success or failure of the authentication.
+
+Simple bind requires the plain text password from the client, so SASL PLAIN will be offered when this identity provider is used.
+
+Later, support will be added for LDAP SASL.
+
+====== External AMQP SASL Provider
+
+Allows the use of an AMQP SASL authentication endpoint that implements the `authServicePlugin` protocol.  When this identity
+provider is used, the Access Control Server acts a SASL proxy.  It will connect to the external provider and relay the
+the AMQP SASL performatives to/from the client.
+
+The response from the External AMQP SASL server must provides an `authenticated-identity`.  It may also provide
+`groups` and may provide `address-authz`.  If the external server provides an `address-authz`, it will be ANDed with the `address-authz` calculated from the `MessagingRoles` and `MessagingRoleBindings`.
+
+The SASL mechanisms offered when this identity provider are in use depend on the mechansims required by the External AMQP SASL server.
+
+====== Internal Identity Provider
+
+The Internal Identity Provider uses the `MessagingUser` resources are a identity/password database.
+
+The password in the MessagingUser record will usually stored salted (defined by  https://tools.ietf.org/html/rfc5802). However
+it will be possible for a request for authentication to arrive before the salting has taken place, to cater for this
+case, the Internal Identity Provider will also accept plain passwords from the database.
+
+This identity provider will offer SASL mechanism PLAIN and SCRAM-SHA-256.
+
+====== Client Auth Identity Provider
+
+For client auth, the identity is established during the TLS negotiation phase between Client and Router.  The router will
+pass the established identity through the `authServicePlugin` to the Access Control Server via the SASL EXTERNAL mechanism
+via the link:https://tools.ietf.org/html/rfc4422#appendix-A[authz-id-string].  (TODO: check Router behaviour).
+
+This identity provider will offer SASL mechanism EXTERNAL.
+
+====== Service Account Identity Provider
+
+For the Service Account provider, the client passes the Service Account's bearer token as a password and a special user-id
+`@@serviceaccount@@`. The Access Control Server will perform a `TokenReview` to check the token's validity and retrieve the user username.  A check is made to ensure that the username refers to a Service Account which is annotated with `enmasse.io/messaging-authentication`.
+
+This identity provider will offer SASL mechanisms PLAIN, XOAUTH2 (Obsolete), link:https://tools.ietf.org/html/rfc7628[OAUTHBEARER]
+
+====== Anonymous Identity Provider 
+
+For the Anonymous Provider, the SASL mechnism ANONYMOUS will be used.
+
+===== Behaviour Without AccessControlService
+
+To support the simple getting started use case, if there are no `AccessControlService` defined in the infra namespace,
```

Comment:
I am not too conformable with this idea.   My main issue is that it violates "secure by default" advocated by OWASP and others.    It also brings with it a complicated switch in behaviour which might confuse the user's mental model and complicate the coding and testing.

I'm wondering if for the getting started user, we should accept that the user has to create an additional three objects:
1. MessagingUser or ServiceAccount
1. MessagingRole
1. MessagingRoleBinding

The content of these objects will be short and self explanatory.  The concepts should already resonate with the user so hopefully the hurdle is not too high.

Reply:
For a getting started experience using anonymous auth, the minimum you'd need is a `AccessControlService` defined as below and a `MessagingRoleBinding`/`MessagingRole` granting `*` address permission to the anonymous subject:

```
anonymous:
       enabled: true
```

As for an out of the box default (with no AccessControlService), I think the system should behave as if the following AccessControlService were in place.

```
internal:
       enabled: true
serviceAccount:
       enabled: true
```

That would mean for a getting started experience with a username/password, the user would just need to create a `MessagingUser` (or `ServiceAccount`) and the `MessagingRoleBinding`/`MessagingRole`.

Reply:
Getting Started Default is now rewritten.